optuna/distributions.py

```diff
@@ -84,14 +84,15 @@ def _asdict(self):
     def __eq__(self, other):
         # type: (Any) -> bool
 
-        if not isinstance(other, type(self)):
-            return False
-
+        if not type(self) is type(other):
+            return NotImplemented
         return self.__dict__ == other.__dict__
 
     def __ne__(self, other):
         # type: (Any) -> bool
 
+        if not type(self) is type(other):
+            return NotImplemented
```

Comment:
We must add this early return/type check in `__ne__` (if we are to define `__ne__` at all. It's not required and as shown here probably shouldn't be defined with Python 3, as it's implicitly defined, in contrast to Python 2 where it's not). `not self.__eq__(other)` casts the results from `self.__eq__(other)` to a `bool` where `bool(NotImplemented) == True` meaning that this method will return `False`. "ne is false" implies that they're equal which should not hold when types are different.

Reply:
[IMO] Since Optuna will drop the support of Python2 in the next release, we can remove the `__ne__` method and use the implicitly defined one instead.

Reply:
That makes sense. I removed the two (and only) `__ne__` definitions.